contracts/DependencyAudit.cdc

```diff
@@ -0,0 +1,127 @@
+                var unstagedDependenciesString = ""
+                var numUnstagedDependencies = unstagedDependencies.length
+                var j = 0
+                while j < numUnstagedDependencies {
+                    if j > 0 {
+                        unstagedDependenciesString = unstagedDependenciesString.concat(", ")
+                    }
+                    unstagedDependenciesString = unstagedDependenciesString.concat(unstagedDependencies[j].toString())
+
+                    j = j + 1
+                }
+
+                // the transactions will fail with a message that looks like this: `error: panic: Found unstaged dependencies: A.0x2ceae959ed1a7e7a.MigrationContractStaging, A.0x2ceae959ed1a7e7a.DependencyAudit`
+                panic("Found unstaged dependencies: ".concat(unstagedDependenciesString))
+            } else {
+                emit UnstagedDependencies(dependencies: unstagedDependencies)
+            }
+        }
+    }
+
+    // The Administrator resorce can be used to add or remove addresses from the excludedAddresses dictionary
+    //
+    access(all) resource Administrator {
+        // addExcludedAddresses add the addresses to the excludedAddresses dictionary
+        access(all) fun addExcludedAddresses(addresses: [Address]) {
+            for address in addresses {
+                DependencyAudit.excludedAddresses[address] = true
+            }
+        }
+
+        // removeExcludedAddresses remove the addresses from the excludedAddresses dictionary
+        access(all) fun removeExcludedAddresses(addresses: [Address]) {
+            for address in addresses {
+                DependencyAudit.excludedAddresses.remove(key: address)
+            }
+        }
+
+        // setPanicOnUnstagedDependencies sets the `panicOnUnstaged` variable to the value of `shouldPanic`
+        access(all) fun setPanicOnUnstagedDependencies(shouldPanic: Bool) {
+            DependencyAudit.panicOnUnstaged = shouldPanic
+            emit PanicOnUnstagedDependenciesChanged(shouldPanic: shouldPanic)
+        }
+
+        // testCheckDependencies is used for testing purposes
+        // It will call the `checkDependencies` function with the provided dependencies
+        // `checkDependencies` is otherwise not callable from the outside
+        access(all) fun testCheckDependencies(_ dependenciesAddresses: [Address], _ dependenciesNames: [String], _ authorizers: [Address]) {
+            return DependencyAudit.checkDependencies(dependenciesAddresses, dependenciesNames, authorizers)
+        }
+    }
+
+    access(all) struct Dependency {
+        access(all) let address: Address
+        access(all) let name: String
+
+        init(address: Address, name: String) {
+            self.address = address
+            self.name = name
+        }
+
+        access(all) fun toString(): String {
+            return "A.".concat(self.address.toString()).concat(".").concat(self.name)
```

Comment:
This resembles contract identifiers, but doesn't exactly match. Contract identifier would include the address without the `0x` prefix - `A.123...789.CONTRACT_NAME` over `A.0x123...789.CONTRACT_NAME`

```suggestion
            let addressString = self.address.toString()
            return "A.".concat(addressString.slice(from: 2, upTo: addressString.length)).concat(".").concat(self.name)
```